src/main/webapp/3Dscript.js

```diff
@@ -0,0 +1,184 @@
+let trajectory;
+let pose;
+let poseTransform = {
+  translateX: 0,
+  translateY: 0,
+  rotate: 0,
+  scale: 0};
+const apiKey = 'AIzaSyDCgKca9sLuoQ9xQDfHUvZf1_KAv06SoTU';
+
+initThreeJs();
+gui();
+animate();
+
+/**
+ * Initializes the essential three.js 3D components then fetchs 
+ * the pose data.
+ */
+function initThreeJs() {
+  scene = new THREE.Scene();
+  camera = new THREE.PerspectiveCamera(
+      /* fov =*/ 75,
+      /* aspectRatio =*/ window.innerWidth / window.innerHeight,
+      /* nearFrustum =*/.1,
+      /* farFrustum =*/ 1000);
+  camera.position.set(0, 1, 1);
+  renderer = new THREE.WebGLRenderer();
+  renderer.setSize(window.innerWidth, window.innerHeight);
+  document.body.appendChild(renderer.domElement);
+  clock = new THREE.Clock();
+  controls = new OrbitControls(camera, renderer.domElement );
+
+  fetchData();
+}
+
+
+/**
+ * This function fetchs pose data from the RunInfo servlet,
+ * it is an asynchronous call requiring addTerrain() to be 
+ * called after the data is fully loaded.
+ */
+function fetchData() {
+    const queryString = window.location.search;
+    const urlParams = new URLSearchParams(queryString);
+    const id = urlParams.get('id');
+    const type = urlParams.get('dataType');
+    fetch('/getrun?id=' + id + '&dataType=' + type)
+    .then(response => response.json())
+    .then(data => pose = data)
+    .then(() => addMap());
+}
+
+/**
+ * Creates the 2D terrain and points a light at it.
+ * It also calls two other functions so it can wait on the fetch 
+ * request and significantly cut down rendering time with 
+ * geometry instancing.
+*/
+function addMap() {
+  // Creates the 2D map.
+  let loader = new THREE.TextureLoader();
+  let material = new THREE.MeshLambertMaterial({ 
+    map: loader.load(
+      'https://maps.googleapis.com/maps/api/staticmap' +
+      '?format=png&center=' + pose[0].lat + ',' + pose[0].lng + 
+      '&zoom=18&size=500x500&key=' + apiKey)
+    });
+  let geometry = new THREE.PlaneGeometry(50,50);
+  const map = new THREE.Mesh(geometry, material);
+  map.position.set(0,-4,0);
+  map.rotation.x = -Math.PI / 2;
+  scene.add(map);
+
+  // Add the light to the scene.
+  let light = new THREE.PointLight(0xffffff, 1, 0 );
+  light.position.set(0, 100, 0);
+  scene.add(light)
+  plotTrajectory();
+  
+  // Adds pose data to the scene as one Geometry instance.
+  material = new THREE.MeshBasicMaterial({color: 'red'});
+  geometry = new THREE.CylinderBufferGeometry(.005, .005, .1);
+  orientation = new THREE.InstancedMesh(geometry, material, pose.length);
+  orientation.instanceMatrix.setUsage(THREE.DynamicDrawUsage);
+  scene.add(orientation);
+  plotOrientation();
+}
+
+/**
+ * This uses the pose data to create a blue line representing the pose
+ * trajectory. 
+ */
+function plotTrajectory() {
+  // Removes any existing trajectory objects for repositioning.
+  scene.remove(trajectory);
+  let coordinates=[];
+  for (point of pose) {
+    coordinates.push(new THREE.Vector3(
```

Comment:
Could you add comment to describe the local frame in three.js? what's the direction of each axes?